pkg/mapper/mapper.go

```diff
@@ -252,53 +256,34 @@ func (m *MetricMapper) InitFromYAMLString(fileContents string, cacheSize int, op
 	return nil
 }
 
-func (m *MetricMapper) InitFromFile(fileName string, cacheSize int, options ...CacheOption) error {
+func (m *MetricMapper) InitFromFile(fileName string) error {
 	mappingStr, err := ioutil.ReadFile(fileName)
 	if err != nil {
 		return err
 	}
 
-	return m.InitFromYAMLString(string(mappingStr), cacheSize, options...)
+	return m.InitFromYAMLString(string(mappingStr))
 }
 
-func (m *MetricMapper) InitCache(cacheSize int, options ...CacheOption) {
-	if cacheSize == 0 {
-		m.cache = NewMetricMapperNoopCache(m.Registerer)
-	} else {
-		o := cacheOptions{
-			cacheType: "lru",
-		}
-		for _, f := range options {
-			f(&o)
-		}
-
-		var (
-			cache MetricMapperCache
-			err   error
-		)
-		switch o.cacheType {
-		case "lru":
-			cache, err = NewMetricMapperCache(m.Registerer, cacheSize)
-		case "random":
-			cache, err = NewMetricMapperRRCache(m.Registerer, cacheSize)
-		default:
-			err = fmt.Errorf("unsupported cache type %q", o.cacheType)
-		}
-
-		if err != nil {
-			log.Fatalf("Unable to setup metric cache. Caused by: %s", err)
-		}
-		m.cache = cache
-	}
+func (m *MetricMapper) UseCache(cache MetricMapperCache) {
+	m.cache = cache
 }
 
 func (m *MetricMapper) GetMapping(statsdMetric string, statsdMetricType MetricType) (*MetricMapping, prometheus.Labels, bool) {
 	m.mutex.RLock()
 	defer m.mutex.RUnlock()
-	result, cached := m.cache.Get(statsdMetric, statsdMetricType)
+
+	// default cache to noop cache if used from an uninitialized mapper
+	if m.cache == nil {
+		m.cache = NewMetricMapperNoopCache()
+	}
```

Comment:
I think we need to take an exclusive lock here, not just an RLock?

Reply:
I guess for performance _and_ correctness we need to check twice? (concurrency is not my strength)

```suggestion
	if m.cache == nil {
		m.mutex.Lock()
		if m.cache == nil {
			m.cache = NewMetricMapperNoopCache()
		}
		m.mutex.Unlock()
	}
```

Reply:
good catch! yeah, we will need to check twice

Reply:
It might make more sense to remove the noop cache entirely, and just hide the part that checks the cache for a result behind the `if m.cache == nil`

Reply:
aaawwwe I liked it (again, aesthetically) but I don't feel strongly about this